File: android/app/src/main/kotlin/com/marrakechguide/ui/components/MapView.kt
package com.marrakechguide.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CloudDownload
import androidx.compose.material.icons.filled.Explore
import androidx.compose.material.icons.filled.LocationOn
import androidx.compose.material.icons.filled.MyLocation
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilledIconButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.marrakechguide.core.model.MapCamera
import com.marrakechguide.core.model.MapCoordinate
import com.marrakechguide.core.model.MapMarker
import com.marrakechguide.core.model.MapRoute
import com.marrakechguide.core.model.MapViewState
import com.marrakechguide.core.model.MarkerCategory
import com.marrakechguide.core.model.TileSource
import com.marrakechguide.ui.theme.Spacing
import kotlin.math.cos
import kotlin.math.sin

/**
 * Offline-capable map view component.
 *
 * This is a placeholder implementation that renders:
 * - A grid background representing the map area
 * - Markers for places
 * - Routes as polylines
 * - User location dot
 *
 * In production, this would be replaced with MapLibre or similar.
 */
@Composable
fun MapView(
    state: MapViewState,
    modifier: Modifier = Modifier,
    tileSource: TileSource = TileSource.PLACEHOLDER,
    onCameraChange: (MapCamera) -> Unit = {},
    onMarkerClick: (MapMarker) -> Unit = {},
    onMyLocationClick: () -> Unit = {},
    onDownloadMapClick: () -> Unit = {}
) {
    var offsetX by remember { mutableFloatStateOf(0f) }
    var offsetY by remember { mutableFloatStateOf(0f) }

    Box(
        modifier = modifier
            .background(Color(0xFFF5F5DC)) // Beige map background
    ) {
        // Map canvas
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .pointerInput(Unit) {
                    detectDragGestures { change, dragAmount ->
                        change.consume()
                        offsetX += dragAmount.x
                        offsetY += dragAmount.y
                        // Update camera based on pan
                        val latDelta = -dragAmount.y / 10000.0
                        val lngDelta = dragAmount.x / 10000.0
                        onCameraChange(
                            state.camera.copy(
                                center = MapCoordinate(
                                    latitude = state.camera.center.latitude + latDelta,
                                    longitude = state.camera.center.longitude + lngDelta
                                )
                            )
                        )
                    }
                }
        ) {
            val centerX = size.width / 2 + offsetX
            val centerY = size.height / 2 + offsetY

            // Draw grid lines (simulating streets)
            val gridColor = Color(0xFFE0DDD0)
            val gridSpacing = 50f
            for (x in 0..size.width.toInt() step gridSpacing.toInt()) {
                drawLine(
                    color = gridColor,
                    start = Offset(x.toFloat(), 0f),
                    end = Offset(x.toFloat(), size.height),
                    strokeWidth = 1f
                )
            }
            for (y in 0..size.height.toInt() step gridSpacing.toInt()) {
                drawLine(
                    color = gridColor,
                    start = Offset(0f, y.toFloat()),
                    end = Offset(size.width, y.toFloat()),
                    strokeWidth = 1f
                )
            }

            // Draw routes
            state.routes.forEach { route ->
                if (route.points.size >= 2) {
                    val path = Path()
                    val firstPoint = coordinateToScreen(
                        route.points[0],
                        state.camera.center,
                        centerX,
                        centerY,
                        state.camera.zoom
                    )
                    path.moveTo(firstPoint.x, firstPoint.y)

                    route.points.drop(1).forEach { point ->
                        val screenPoint = coordinateToScreen(
                            point,
                            state.camera.center,
                            centerX,
                            centerY,
                            state.camera.zoom
                        )
                        path.lineTo(screenPoint.x, screenPoint.y)
                    }

                    drawPath(
                        path = path,
                        color = Color(0xFF4A90D9),
                        style = Stroke(width = route.width * 2)
                    )
                }
            }

            // Draw markers
            state.markers.forEach { marker ->
                val screenPos = coordinateToScreen(
                    marker.coordinate,
                    state.camera.center,
                    centerX,
                    centerY,
                    state.camera.zoom
                )

                val markerColor = when (marker.category) {
                    MarkerCategory.HOME_BASE -> Color(0xFF4CAF50) // Green
                    MarkerCategory.ROUTE_STOP -> Color(0xFF2196F3) // Blue
                    MarkerCategory.RESTAURANT, MarkerCategory.CAFE -> Color(0xFFFF9800) // Orange
                    MarkerCategory.LANDMARK, MarkerCategory.MUSEUM -> Color(0xFF9C27B0) // Purple
                    else -> Color(0xFFE53935) // Red
                }

                // Draw marker pin
                drawCircle(
                    color = markerColor,
                    radius = if (marker.isSelected) 16f else 12f,
                    center = screenPos
                )
                drawCircle(
                    color = Color.White,
                    radius = if (marker.isSelected) 8f else 6f,
                    center = screenPos
                )
            }

            // Draw user location
            state.userLocation?.let { location ->
                val userPos = coordinateToScreen(
                    location,
                    state.camera.center,
                    centerX,
                    centerY,
                    state.camera.zoom
                )

                // Accuracy circle
                drawCircle(
                    color = Color(0x334A90D9),
                    radius = 30f,
                    center = userPos
                )

                // User dot
                drawCircle(
                    color = Color(0xFF4A90D9),
                    radius = 10f,
                    center = userPos
                )
                drawCircle(
                    color = Color.White,
                    radius = 6f,
                    center = userPos
                )

                // Heading indicator
                state.userHeading?.let { heading ->
                    rotate(heading.toFloat(), userPos) {
                        val trianglePath = Path().apply {
                            moveTo(userPos.x, userPos.y - 20f)
                            lineTo(userPos.x - 6f, userPos.y - 8f)
                            lineTo(userPos.x + 6f, userPos.y - 8f)
                            close()
                        }
                        drawPath(
                            path = trianglePath,
                            color = Color(0xFF4A90D9)
                        )
                    }
                }
            }
        }

        // Loading indicator
        if (state.isLoading) {
            CircularProgressIndicator(
                modifier = Modifier
                    .align(Alignment.Center)
                    .size(48.dp)
            )
        }

        // Offline map prompt
        if (tileSource == TileSource.PLACEHOLDER && !state.isLoading) {
            OfflineMapPrompt(
                onDownloadClick = onDownloadMapClick,
                modifier = Modifier
                    .align(Alignment.TopCenter)
                    .padding(top = Spacing.lg)
            )
        }

        // Map controls
        Column(
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(Spacing.md)
        ) {
            // My location button
            FilledIconButton(
                onClick = onMyLocationClick,
                colors = IconButtonDefaults.filledIconButtonColors(
                    containerColor = MaterialTheme.colorScheme.surface
                )
            ) {
                Icon(
                    Icons.Default.MyLocation,
                    contentDescription = "My location",
                    tint = MaterialTheme.colorScheme.primary
                )
            }
        }

        // Error message
        state.error?.let { error ->
            Card(
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .padding(Spacing.md),
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.errorContainer
                )
            ) {
                Text(
                    text = error,
                    modifier = Modifier.padding(Spacing.sm),
                    color = MaterialTheme.colorScheme.onErrorContainer
                )
            }
        }
    }
}

@Composable
private fun OfflineMapPrompt(
    onDownloadClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Card(
        modifier = modifier,
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.primaryContainer
        ),
        onClick = onDownloadClick
    ) {
        Row(
            modifier = Modifier.padding(Spacing.md),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                Icons.Default.CloudDownload,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary
            )
            Spacer(Modifier.width(Spacing.sm))
            Column {
                Text(
                    text = "Offline Maps Available",
                    style = MaterialTheme.typography.labelLarge,
                    fontWeight = FontWeight.SemiBold
                )
                Text(
                    text = "Download for navigation without internet",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

/**
 * Convert a geographic coordinate to screen position.
 */
private fun coordinateToScreen(
    coordinate: MapCoordinate,
    center: MapCoordinate,
    centerX: Float,
    centerY: Float,
    zoom: Double
): Offset {
    val scale = zoom * 10000
    val x = centerX + ((coordinate.longitude - center.longitude) * scale).toFloat()
    val y = centerY - ((coordinate.latitude - center.latitude) * scale).toFloat()
    return Offset(x, y)
}

/**
 * Simple compass indicator component.
 */
@Composable
fun CompassIndicator(
    bearing: Double,
    modifier: Modifier = Modifier
) {
    Surface(
        modifier = modifier.size(40.dp),
        shape = CircleShape,
        color = MaterialTheme.colorScheme.surface,
        shadowElevation = 2.dp
    ) {
        Box(contentAlignment = Alignment.Center) {
            Canvas(modifier = Modifier.size(32.dp)) {
                rotate(-bearing.toFloat()) {
                    // North indicator
                    drawLine(
                        color = Color.Red,
                        start = Offset(center.x, center.y - 12f),
                        end = Offset(center.x, center.y + 4f),
                        strokeWidth = 3f
                    )
                    // South indicator
                    drawLine(
                        color = Color.Gray,
                        start = Offset(center.x, center.y + 4f),
                        end = Offset(center.x, center.y + 12f),
                        strokeWidth = 3f
                    )
                }
            }
        }
    }
}

File: ios/MarrakechGuide/MarrakechGuide/Shared/Components/MapView.swift
import SwiftUI
import CoreLocation

/// Offline-capable map view component.
///
/// This is a placeholder implementation that renders:
/// - A grid background representing the map area
/// - Markers for places
/// - Routes as polylines
/// - User location dot
///
/// In production, this would be replaced with MapLibre or MapKit.
struct OfflineMapView: View {
    let state: MapViewState
    let tileSource: TileSource
    let onCameraChange: (MapCamera) -> Void
    let onMarkerClick: (MapMarker) -> Void
    let onMyLocationClick: () -> Void
    let onDownloadMapClick: () -> Void

    @State private var offset: CGSize = .zero
    @GestureState private var dragOffset: CGSize = .zero

    init(
        state: MapViewState,
        tileSource: TileSource = .placeholder,
        onCameraChange: @escaping (MapCamera) -> Void = { _ in },
        onMarkerClick: @escaping (MapMarker) -> Void = { _ in },
        onMyLocationClick: @escaping () -> Void = {},
        onDownloadMapClick: @escaping () -> Void = {}
    ) {
        self.state = state
        self.tileSource = tileSource
        self.onCameraChange = onCameraChange
        self.onMarkerClick = onMarkerClick
        self.onMyLocationClick = onMyLocationClick
        self.onDownloadMapClick = onDownloadMapClick
    }

    var body: some View {
        ZStack {
            // Map background
            Color(red: 0.96, green: 0.96, blue: 0.86)

            // Map canvas
            GeometryReader { geometry in
                Canvas { context, size in
                    let centerX = size.width / 2 + offset.width + dragOffset.width
                    let centerY = size.height / 2 + offset.height + dragOffset.height

                    // Draw grid lines (simulating streets)
                    let gridColor = Color(red: 0.88, green: 0.87, blue: 0.82)
                    let gridSpacing: CGFloat = 50

                    for x in stride(from: 0, to: size.width, by: gridSpacing) {
                        context.stroke(
                            Path { path in
                                path.move(to: CGPoint(x: x, y: 0))
                                path.addLine(to: CGPoint(x: x, y: size.height))
                            },
                            with: .color(gridColor),
                            lineWidth: 1
                        )
                    }
                    for y in stride(from: 0, to: size.height, by: gridSpacing) {
                        context.stroke(
                            Path { path in
                                path.move(to: CGPoint(x: 0, y: y))
                                path.addLine(to: CGPoint(x: size.width, y: y))
                            },
                            with: .color(gridColor),
                            lineWidth: 1
                        )
                    }

                    // Draw routes
                    for route in state.routes where route.points.count >= 2 {
                        var path = Path()
                        let firstPoint = coordinateToScreen(
                            route.points[0],
                            center: state.camera.center,
                            centerX: centerX,
                            centerY: centerY,
                            zoom: state.camera.zoom
                        )
                        path.move(to: firstPoint)

                        for point in route.points.dropFirst() {
                            let screenPoint = coordinateToScreen(
                                point,
                                center: state.camera.center,
                                centerX: centerX,
                                centerY: centerY,
                                zoom: state.camera.zoom
                            )
                            path.addLine(to: screenPoint)
                        }

                        context.stroke(
                            path,
                            with: .color(Color(red: 0.29, green: 0.56, blue: 0.85)),
                            lineWidth: CGFloat(route.width * 2)
                        )
                    }

                    // Draw markers
                    for marker in state.markers {
                        let screenPos = coordinateToScreen(
                            marker.coordinate,
                            center: state.camera.center,
                            centerX: centerX,
                            centerY: centerY,
                            zoom: state.camera.zoom
                        )

                        let markerColor = color(for: marker.category)
                        let radius: CGFloat = marker.isSelected ? 16 : 12

                        // Marker pin
                        context.fill(
                            Path(ellipseIn: CGRect(
                                x: screenPos.x - radius,
                                y: screenPos.y - radius,
                                width: radius * 2,
                                height: radius * 2
                            )),
                            with: .color(markerColor)
                        )

                        // White center
                        let innerRadius = marker.isSelected ? 8.0 : 6.0
                        context.fill(
                            Path(ellipseIn: CGRect(
                                x: screenPos.x - innerRadius,
                                y: screenPos.y - innerRadius,
                                width: innerRadius * 2,
                                height: innerRadius * 2
                            )),
                            with: .color(.white)
                        )
                    }

                    // Draw user location
                    if let location = state.userLocation {
                        let userPos = coordinateToScreen(
                            location,
                            center: state.camera.center,
                            centerX: centerX,
                            centerY: centerY,
                            zoom: state.camera.zoom
                        )

                        // Accuracy circle
                        context.fill(
                            Path(ellipseIn: CGRect(
                                x: userPos.x - 30,
                                y: userPos.y - 30,
                                width: 60,
                                height: 60
                            )),
                            with: .color(Color(red: 0.29, green: 0.56, blue: 0.85).opacity(0.2))
                        )

                        // User dot
                        context.fill(
                            Path(ellipseIn: CGRect(
                                x: userPos.x - 10,
                                y: userPos.y - 10,
                                width: 20,
                                height: 20
                            )),
                            with: .color(Color(red: 0.29, green: 0.56, blue: 0.85))
                        )

                        // White center
                        context.fill(
                            Path(ellipseIn: CGRect(
                                x: userPos.x - 6,
                                y: userPos.y - 6,
                                width: 12,
                                height: 12
                            )),
                            with: .color(.white)
                        )

                        // Heading indicator
                        if let heading = state.userHeading {
                            let radians = heading * .pi / 180
                            var trianglePath = Path()
                            trianglePath.move(to: CGPoint(x: userPos.x, y: userPos.y - 20))
                            trianglePath.addLine(to: CGPoint(x: userPos.x - 6, y: userPos.y - 8))
                            trianglePath.addLine(to: CGPoint(x: userPos.x + 6, y: userPos.y - 8))
                            trianglePath.closeSubpath()

                            context.drawLayer { ctx in
                                ctx.translateBy(x: userPos.x, y: userPos.y)
                                ctx.rotate(by: .radians(radians))
                                ctx.translateBy(x: -userPos.x, y: -userPos.y)
                                ctx.fill(trianglePath, with: .color(Color(red: 0.29, green: 0.56, blue: 0.85)))
                            }
                        }
                    }
                }
                .gesture(
                    DragGesture()
                        .updating($dragOffset) { value, state, _ in
                            state = value.translation
                        }
                        .onEnded { value in
                            offset = CGSize(
                                width: offset.width + value.translation.width,
                                height: offset.height + value.translation.height
                            )
                            // Update camera
                            let latDelta = -value.translation.height / 10000.0
                            let lngDelta = value.translation.width / 10000.0
                            onCameraChange(MapCamera(
                                center: MapCoordinate(
                                    latitude: state.camera.center.latitude + latDelta,
                                    longitude: state.camera.center.longitude + lngDelta
                                ),
                                zoom: state.camera.zoom,
                                bearing: state.camera.bearing,
                                tilt: state.camera.tilt
                            ))
                        }
                )
            }

            // Loading indicator
            if state.isLoading {
                ProgressView()
                    .progressViewStyle(CircularProgressViewStyle())
                    .scaleEffect(1.5)
            }

            // Offline map prompt
            if tileSource == .placeholder && !state.isLoading {
                VStack {
                    OfflineMapPrompt(onDownloadClick: onDownloadMapClick)
                        .padding(.top, Theme.Spacing.lg)
                    Spacer()
                }
            }

            // Map controls
            VStack {
                Spacer()
                HStack {
                    Spacer()
                    Button(action: onMyLocationClick) {
                        Image(systemName: "location.fill")
                            .font(.title3)
                            .foregroundStyle(Theme.Adaptive.primary)
                            .frame(width: 44, height: 44)
                            .background(Theme.Adaptive.cardBackground)
                            .clipShape(Circle())
                            .shadow(radius: 2)
                    }
                    .padding(Theme.Spacing.md)
                }
            }

            // Error message
            if let error = state.error {
                VStack {
                    Spacer()
                    Text(error)
                        .font(.themeCaption)
                        .foregroundStyle(.white)
                        .padding(Theme.Spacing.sm)
                        .background(Color.red)
                        .cornerRadius(Theme.CornerRadius.small)
                        .padding(Theme.Spacing.md)
                }
            }
        }
    }

    private func coordinateToScreen(
        _ coordinate: MapCoordinate,
        center: MapCoordinate,
        centerX: CGFloat,
        centerY: CGFloat,
        zoom: Double
    ) -> CGPoint {
        let scale = zoom * 10000
        let x = centerX + CGFloat((coordinate.longitude - center.longitude) * scale)
        let y = centerY - CGFloat((coordinate.latitude - center.latitude) * scale)
        return CGPoint(x: x, y: y)
    }

    private func color(for category: MarkerCategory) -> Color {
        switch category {
        case .homeBase:
            return Color(red: 0.30, green: 0.69, blue: 0.31) // Green
        case .routeStop:
            return Color(red: 0.13, green: 0.59, blue: 0.95) // Blue
        case .restaurant, .cafe:
            return Color(red: 1.0, green: 0.60, blue: 0.0) // Orange
        case .landmark, .museum:
            return Color(red: 0.61, green: 0.15, blue: 0.69) // Purple
        default:
            return Color(red: 0.90, green: 0.22, blue: 0.21) // Red
        }
    }
}

// MARK: - Offline Map Prompt

private struct OfflineMapPrompt: View {
    let onDownloadClick: () -> Void

    var body: some View {
        Button(action: onDownloadClick) {
            HStack(spacing: Theme.Spacing.sm) {
                Image(systemName: "icloud.and.arrow.down")
                    .font(.title3)
                    .foregroundStyle(Theme.Adaptive.primary)

                VStack(alignment: .leading, spacing: 2) {
                    Text("Offline Maps Available")
                        .font(.themeSubheadline.weight(.semibold))
                        .foregroundStyle(Theme.Adaptive.textPrimary)
                    Text("Download for navigation without internet")
                        .font(.themeCaption)
                        .foregroundStyle(Theme.Adaptive.textSecondary)
                }
            }
            .padding(Theme.Spacing.md)
            .background(Theme.Adaptive.primaryContainer)
            .cornerRadius(Theme.CornerRadius.medium)
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Compass Indicator

/// Simple compass indicator component.
struct CompassIndicator: View {
    let bearing: Double

    var body: some View {
        ZStack {
            Circle()
                .fill(Theme.Adaptive.cardBackground)
                .frame(width: 40, height: 40)
                .shadow(radius: 2)

            Canvas { context, size in
                let center = CGPoint(x: size.width / 2, y: size.height / 2)

                context.drawLayer { ctx in
                    ctx.translateBy(x: center.x, y: center.y)
                    ctx.rotate(by: .degrees(-bearing))
                    ctx.translateBy(x: -center.x, y: -center.y)

                    // North indicator (red)
                    ctx.stroke(
                        Path { path in
                            path.move(to: CGPoint(x: center.x, y: center.y - 12))
                            path.addLine(to: CGPoint(x: center.x, y: center.y + 4))
                        },
                        with: .color(.red),
                        lineWidth: 3
                    )

                    // South indicator (gray)
                    ctx.stroke(
                        Path { path in
                            path.move(to: CGPoint(x: center.x, y: center.y + 4))
                            path.addLine(to: CGPoint(x: center.x, y: center.y + 12))
                        },
                        with: .color(.gray),
                        lineWidth: 3
                    )
                }
            }
        }
        .frame(width: 40, height: 40)
    }
}

// MARK: - Preview

#Preview {
    OfflineMapView(
        state: MapViewState(
            markers: [
                MapMarker(
                    id: "1",
                    coordinate: .medinaCenter,
                    title: "Jemaa el-Fna",
                    category: .landmark
                )
            ],
            userLocation: MapCoordinate(latitude: 31.6260, longitude: -7.9895),
            userHeading: 45
        )
    )
}
